### PR TITLE
[CWS] add a approver mode to bypass discarders

### PR DESCRIPTION
**Pull request by safchain**
### What does this PR do?

Add an Approver mode which can be used in capabilities declaration in order to remove the rule having an approver with this mode from the set of rule used to generate discarders. The goal is to be able to write approver only rules which have in-kernel filters but which would avoid the discarder detection.

ex: 

```
open.file.path ~= "/etc/*" # parent discarder
open.file.name != "" && process.auid == 1000 # no more `open.file.path` discarder except if we set `process.auid` as "approver only" 
```

This PR is part of a series which aims to improve the approver mechanism. There is no functional impact, it is only add unit tests.

### CODE REVIEW

File: pkg/security/probe/discarders_linux.go [hunk 1]
```diff
@@ -332,8 +332,11 @@ func (id *inodeDiscarders) getParentDiscarderFnc(rs *rules.RuleSet, eventType mo
 			return false, false, err
 		}
 
-		if isDiscarder, _ := rules.IsDiscarder(id.evalCtx, basenameField, basenameRules); !isDiscarder {
-			return false, true, nil
+		if len(basenameRules) > 0 {
+			if isDiscarder, err := rules.IsDiscarder(id.evalCtx, basenameField, basenameRules); !isDiscarder {
+				fmt.Printf("COM: %v %v\n", isDiscarder, err)
```
Comment (paulcacheux) [inline pkg/security/probe/discarders_linux.go:337]: Is this expected ? The printf

File: cmd/security-agent/subcommands/runtime/command_test.go [hunk 2]
```diff
@@ -133,19 +133,22 @@ func Test_checkPoliciesLoaded(t *testing.T) {
 					{
 						"Field": "open.file.path",
 						"Value": "/etc/gshadow",
-						"Type": "scalar"
+						"Type": "scalar",
+						"Mode": 0
```
Comment (paulcacheux) [inline cmd/security-agent/subcommands/runtime/command_test.go:137]: Why not use `NormalMode` here ?

File: pkg/security/secl/rules/ruleset.go [hunk 3]
```diff
@@ -603,13 +604,22 @@ func (rs *RuleSet) GetFieldValues(field eval.Field) []eval.FieldValue {
 
 // IsDiscarder partially evaluates an Event against a field
 func IsDiscarder(ctx *eval.Context, field eval.Field, rules []*Rule) (bool, error) {
+	var isDiscarder bool
+
 	for _, rule := range rules {
+		// ignore rule that can't generate discarders
+		if rule.NoDiscarder {
```
Comment (paulcacheux) [inline pkg/security/secl/rules/ruleset.go:611]: Doesn't this mean that we might miss events for `NoDiscarder` rules ? I'm surprised we don't need to check the partial evaluation here.

File: pkg/security/probe/discarders_linux.go (hunk 1 above)
Comment (safchain) [reply pkg/security/probe/discarders_linux.go:337]: oups, removed

**Review by paulcacheux — APPROVED** [review]

File: pkg/security/secl/rules/ruleset.go (hunk 3 above)
Comment (safchain) [reply pkg/security/secl/rules/ruleset.go:611]: Discussed about it offline. The NoDiscarder flag will be applied for rule having specific approver that will supplant the discarder mechanism kernel side.

**Review by gjulianm — APPROVED** [review]

**Pull request merged**